editor/plugins/gradient_editor_plugin.cpp

```diff
@@ -31,27 +31,627 @@
+			draw_rect(Rect2(handle_start_x + handle_thickness / 2.0, h * 0.9 - handle_thickness / 2.0, handle_width - handle_thickness, h * 0.1), inside_col_opaque, true);
+		} else {
+			draw_rect(rect, inside_col, true);
+		}
+
+		if (selected_index == i) {
+			// Handle is selected.
+			draw_rect(rect, border_col, false, handle_thickness);
+			draw_line(Vector2(handle_x_pos, 0), Vector2(handle_x_pos, h / 2 - handle_thickness), border_col, handle_thickness);
+			if (inside_col.a < 1) {
+				draw_line(Vector2(handle_start_x + handle_thickness / 2.0, h * 0.9 - handle_thickness), Vector2(handle_start_x + handle_width - handle_thickness / 2.0, h * 0.9 - handle_thickness), border_col, handle_thickness);
+			}
+			rect = rect.grow(-handle_thickness);
+			const Color focus_col = get_theme_color(SNAME("accent_color"), EditorStringName(Editor));
+			draw_rect(rect, has_focus() ? focus_col : focus_col.darkened(0.4), false, handle_thickness);
+			rect = rect.grow(-handle_thickness);
+			draw_rect(rect, border_col, false, handle_thickness);
+		} else {
+			// Handle isn't selected.
+			border_col.a = 0.9;
+			draw_rect(rect, border_col, false, handle_thickness);
+			draw_line(Vector2(handle_x_pos, 0), Vector2(handle_x_pos, h / 2 - handle_thickness), border_col, handle_thickness);
+			if (inside_col.a < 1) {
+				draw_line(Vector2(handle_start_x + handle_thickness / 2.0, h * 0.9 - handle_thickness), Vector2(handle_start_x + handle_width - handle_thickness / 2.0, h * 0.9 - handle_thickness), border_col, handle_thickness);
+			}
+			if (hovered_index == i) {
+				// Draw a subtle translucent rect inside the handle if it's being hovered.
+				rect = rect.grow(-handle_thickness);
+				border_col.a = 0.54;
+				draw_rect(rect, border_col, false, handle_thickness);
+			}
+		}
+	}
+
+	// Draw "button" for color selector.
+	int button_offset = total_w + handle_width + draw_spacing;
+	if (selected_index != -1) {
+		Color grabbed_col = gradient->get_color(selected_index);
+		if (grabbed_col.a < 1) {
+			draw_texture_rect(get_editor_theme_icon(SNAME("GuiMiniCheckerboard")), Rect2(button_offset, 0, h, h), true);
+		}
+		draw_rect(Rect2(button_offset, 0, h, h), grabbed_col);
+		if (grabbed_col.r > 1 || grabbed_col.g > 1 || grabbed_col.b > 1) {
+			// Draw an indicator to denote that the currently selected color is "overbright".
+			draw_texture(get_theme_icon(SNAME("overbright_indicator"), SNAME("ColorPicker")), Point2(button_offset, 0));
+		}
+	} else {
+		// If no color is selected, draw grey color with 'X' on top.
+		draw_rect(Rect2(button_offset, 0, h, h), Color(0.5, 0.5, 0.5, 1));
+		draw_line(Vector2(button_offset, 0), Vector2(button_offset + h, h), Color(0.8, 0.8, 0.8));
+		draw_line(Vector2(button_offset, h), Vector2(button_offset + h, 0), Color(0.8, 0.8, 0.8));
+	}
+}
+
+void GradientEdit::_notification(int p_what) {
 	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+			picker->connect("color_changed", callable_mp(this, &GradientEdit::_color_changed));
```

Comment:
This should be in the constructor, or is there a reason for this approach?

Reply:
Not sure, also copied code from the old class. Maybe signals connected in a constructor are kinda inconsistent? I remember something being more inconsistent in constructors than enter_tree, but probably not this, so I'll move this.

Reply:
You can't use themes in the constructor, and you also may have issues interacting with the node being constructed itself. But connecting to signals of children should be completely fine.